frontend/views/pages/Contributions.vue

```diff
@@ -208,28 +187,79 @@ export default {
       }
     }
   },
-  validations: {
-    form: {
-      incomeAmount: {
-        [L('field is required')]: required,
-        [L('cannot be negative')]: v => v >= 0,
-        [L('cannot have more than 2 decimals')]: decimals(2)
-      },
-      pledgeAmount: { required, minValue: v => v >= 0, decimals: decimals(2) }
-    }
-  },
   computed: {
     ...mapGetters([
+      'ourUsername',
       'groupSettings',
+      'groupMembersCount',
       'groupProfile',
+      'groupProfiles',
+      'thisMonthsPayments',
       'groupMincomeFormatted',
-      'ourUsername'
+      'groupIncomeDistribution',
+      'globalProfile'
     ]),
-    doesReceiveMonetary () {
-      return !!this.fakeStore.receiving.monetary && !this.ephemeral.isEditingIncome
+    memberGroupProfile () {
+      return this.groupProfile(this.ourUsername) || {}
+    },
+    upTo () {
+      const amount = this.memberGroupProfile[this.memberGroupProfile.incomeDetailsType]
+      if (!amount) return false
+      return this.currency.displayWithCurrency(this.needsIncome ? this.groupSettings.mincomeAmount - amount : amount)
+    },
+    doesReceive () {
+      return this.needsIncome || this.receivingNonMonetary
+    },
+    doesGive () {
+      return this.hasPayments || this.memberGroupProfile.nonMonetaryContributions
+    },
+    needsIncome () {
+      return this.memberGroupProfile.incomeDetailsType === 'incomeAmount'
+    },
+    hasPayments () {
+      return Object.keys(this.givingMonetary.list).length > 0
+    },
+    currency () {
+      return currencies[this.groupSettings.mincomeCurrency]
+    },
+    distribution () {
+      return this.thisMonthsPayments.frozenDistribution || this.groupIncomeDistribution
+    },
+    receivingNonMonetary () {
+      const groupProfiles = this.groupProfiles
+      return Object.keys(groupProfiles)
+        .filter(key => key !== this.ourUsername && groupProfiles[key].nonMonetaryContributions)
+        .reduce((list, username) => {
+          const nonMonetary = groupProfiles[username].nonMonetaryContributions
+          nonMonetary.forEach((what) => {
+            const contributionIndex = list.findIndex(x => x.what === what)
+            const displayName = this.displayName(username)
+            contributionIndex >= 0
+              ? list[contributionIndex].who.push(displayName)
+              : list.push({ who: [displayName], what })
+          })
+          return list
+        }, [])
     },
-    doesGiveMonetary () {
-      return !!this.fakeStore.giving.monetary && !this.ephemeral.isEditingIncome
+    receivingMonetary () {
+      const distributionTo = this.distribution.filter(p => p.to === this.ourUsername)
+      let total = distributionTo.reduce((acc, payment) => acc + payment.amount, 0)
+      const list = distributionTo.reduce((acc, payment) => this.displayName(payment.from), [])
```

Comment:
Here, you can do one loop instead of two:

```js
const list = []
const total = this.currency.displayWithCurrency(
  distributionTo.reduce((acc, payment) => {
    list.push(this.displayName(payment.from))
    return acc + payment.amount
  }, 0)
)
```